modules/ingestor/src/service/advisory/csaf/creator.rs

```diff
@@ -0,0 +1,147 @@
+use crate::{
+    graph::{
+        advisory::advisory_vulnerability::{VersionInfo, VersionSpec},
+        purl::creator::PurlCreator,
+    },
+    service::{advisory::csaf::util::resolve_purls, Error},
+};
+use csaf::{definitions::ProductIdT, Csaf};
+use sea_orm::{ActiveValue::Set, ColumnTrait, ConnectionTrait, EntityTrait, QueryFilter};
+use sea_query::IntoCondition;
+use std::{collections::hash_map::Entry, collections::HashMap};
+use trustify_common::{db::chunk::EntityChunkedIter, purl::Purl};
+use trustify_entity::{package_status, status, version_range};
+use uuid::Uuid;
+
+struct PackageStatus {
+    package: Purl,
+    status: &'static str,
+    info: VersionInfo,
+}
+
+pub struct PackageStatusCreator {
+    advisory_id: Uuid,
+    vulnerability_id: i32,
+    entries: Vec<PackageStatus>,
+}
+
+impl PackageStatusCreator {
+    pub fn new(advisory_id: Uuid, vulnerability_id: i32) -> Self {
+        Self {
+            advisory_id,
+            vulnerability_id,
+            entries: Vec::new(),
+        }
+    }
+
+    pub fn add_all(&mut self, csaf: &Csaf, ps: &Option<Vec<ProductIdT>>, status: &'static str) {
+        for r in ps.iter().flatten() {
+            for purl in resolve_purls(csaf, r) {
+                let package = Purl::from(purl.clone());
+
+                if let Some(version) = package.version.clone() {
+                    self.entries.push(PackageStatus {
+                        package,
+                        status,
+                        info: VersionInfo {
+                            scheme: "generic".to_string(),
```

Comment:
not your problem, just noting: there's a few schemes which happen to match the purl `type` so we could avoid `generic` perhaps, if we know it's a python, or maven, or rust (semver) type.